analysis/complex.lean

```diff
@@ -0,0 +1,281 @@
+λ z,  { re := z.re / norm_squared z,
+        im := -z.im / norm_squared z }
+
+instance : has_add complex := ⟨complex.add⟩
+instance : has_neg complex := ⟨complex.neg⟩
+instance : has_sub complex := ⟨λx y, x + - y⟩
+instance : has_mul complex := ⟨complex.mul⟩
+instance : has_inv complex := ⟨complex.inv⟩
+instance : has_div complex := ⟨λx y, x * y⁻¹⟩
+
+-- I was initially astounded to find that at some point there was a typo in has_div but
+-- this didn't cause any problems at all. I have since understood what is
+-- going on: "/" is never used in the field axioms, only ^{-1} .
+
+-- These are very useful for proofs in the library so I make them local simp lemmas.
+
+lemma proj_zero_re : (0:complex).re=0 := rfl
+lemma proj_zero_im : (0:complex).im=0 := rfl
+lemma proj_one_re : (1:complex).re=1 := rfl
+lemma proj_one_im : (1:complex).im=0 := rfl
+lemma proj_i_re : complex.i.re=0 := rfl
+lemma proj_i_im : complex.i.im=1 := rfl
+lemma proj_conj_re (z : complex) : (conjugate z).re = z.re := rfl
+lemma proj_conj_im (z : complex) : (conjugate z).im = -z.im := rfl
+lemma proj_add_re (z w: complex) : (z+w).re=z.re+w.re := rfl
+lemma proj_add_im (z w: complex) : (z+w).im=z.im+w.im := rfl
+lemma proj_neg_re (z: complex) : (-z).re=-z.re := rfl
+lemma proj_neg_im (z: complex) : (-z).im=-z.im := rfl
+lemma proj_neg_re' (z: complex) : (neg z).re=-z.re := rfl
+lemma proj_neg_im' (z: complex) : (neg z).im=-z.im := rfl
+lemma proj_sub_re (z w : complex) : (z-w).re=z.re-w.re := rfl
+lemma proj_sub_im (z w : complex) : (z-w).im=z.im-w.im := rfl
+lemma proj_mul_re (z w: complex) : (z*w).re=z.re*w.re-z.im*w.im := rfl
+lemma proj_mul_im (z w: complex) : (z*w).im=z.re*w.im+z.im*w.re := rfl
+lemma proj_of_real_re (r:real) : (of_real r).re = r := rfl
+lemma proj_of_real_im (r:real) : (of_real r).im = 0 := rfl
+local attribute [simp] proj_zero_re proj_zero_im proj_one_re proj_one_im
+local attribute [simp] proj_i_re proj_i_im proj_conj_re proj_conj_im
+local attribute [simp] proj_add_re proj_add_im proj_neg_re proj_neg_im
+local attribute [simp] proj_neg_re' proj_neg_im' proj_sub_re proj_sub_im
+local attribute [simp] proj_mul_re proj_mul_im proj_of_real_re proj_of_real_im
+
+lemma norm_squared_pos_of_nonzero (z : complex) (H : z ≠ 0) : norm_squared z > 0 :=
+begin -- far more painful than it should be but I need it for inverses
+suffices : z.re ≠ 0 ∨ z.im ≠ 0,
+  apply lt_of_le_of_ne,
+    exact add_nonneg (mul_self_nonneg _) (mul_self_nonneg _),
+  intro H2,
+  cases this with Hre Him,
+    exact Hre (eq_zero_of_mul_self_add_mul_self_eq_zero (eq.symm H2)),
+  unfold norm_squared at H2,rw [add_comm] at H2,
+  exact Him (eq_zero_of_mul_self_add_mul_self_eq_zero (eq.symm H2)),
+have : ¬ (z.re = 0 ∧ z.im = 0),
+  intro H2,
+  exact H (eq_of_re_eq_and_im_eq z 0 H2),
+cases classical.em (z.re = 0) with Hre_eq Hre_ne,
+  right,
+  intro H2,
+  apply this,
+  exact ⟨Hre_eq,H2⟩,
+left,assumption,
+end
+
+-- I don't know how to set up
+-- real.cast_zero etc
```

Comment:
It's true that you can define a `real.cast`, but it has complicated topological side conditions and I'd rather not define it until I need it. Using `of_real` as the designated coercion is fine, although we will want theorems saying that the casts commute with `of_real`, i.e. `of_real ((n : nat) : real) = ((n : nat) : complex)` and similarly for rat and int.

Reply:
I wrote these for nat and int; I almost wrote one for rat but I needed that the complexes had characteristic zero. I proved this but then for some reason I could not put it all together -- I was getting timeouts. I've left my proof in but commented it out.